# 2. Data Preparation

In this notebook we aim to do two things with the raw data `data/all_car_adverts.csv`.

1. Clean the data by fixing the issues we identified in our initial exploration.
2. Transform some of the columns such that they can be used in a machine learning model, i.e. Feature Engineering.


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
df = pd.read_csv("../data/all_car_adverts.csv")
df.head()

,Unnamed: 0,make,model,variant,car_price,car_badges,car_title,car_sub_title,car_attention_grabber,car_specs,...,num_owner,ulez,full_service,part_service,part_warranty,full_dealership,first_year_road_tax,brand_new,finance_available,discounted
0,0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",...,5.0,0,0,0,0,0,0,0,0,0
1,1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",...,NaN,0,0,0,0,0,0,0,0,0
2,2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",...,3.0,0,0,0,0,0,0,0,0,0
3,3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",...,NaN,0,0,0,0,0,0,0,0,0
4,4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",...,NaN,0,0,0,0,0,0,0,0,0


## 2.1. Column Renaming

In this section we clean the dataset and fix as many of the issues that we identified during initial analysis as possible.

#### Redundant Index Column
Firstly, we identified `Unnamed: 0` as an index column. This is a redundant column since we have access to indices through pandas' API anyway.

In [3]:
df = df.drop(columns=["Unnamed: 0"])
df.head()

,make,model,variant,car_price,car_badges,car_title,car_sub_title,car_attention_grabber,car_specs,car_seller,...,num_owner,ulez,full_service,part_service,part_warranty,full_dealership,first_year_road_tax,brand_new,finance_available,discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


#### Column Naming

Due to personal preference, we also change the naming convention of the columns from `snake_case` to `CamelCase`.

In [4]:
renaming_map = {
    x: x.replace("_", " ").title().replace(" ", "")
    for x in df.columns.values
}
df = df.rename(columns=renaming_map)

df.head()

,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,Ulez,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


We rename `Ulez` to `ULEZ` because it is an acronym and we also rename `FeulType` to `FuelType` to fix the typo.

In [5]:
renaming_map = {
    "Ulez": "ULEZ",
    "FeulType": "FuelType"
}
df = df.rename(columns=renaming_map)

df.head()

,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,0,0,0,0,0,0,0,0,0
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,0,0,0,0,0,0,0,0,0
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,0,0,0,0,0,0,0,0,0
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,0,0,0,0,0,0,0,0,0


## 2.2. Data Types

We notice that some of the data types seem incorrect. For example, `NumOwner` is clearly not an integer column - but cars can only have an integer amount of owners.

In [6]:
df.dtypes

Make                       str
Model                      str
Variant                    str
CarPrice               float64
CarBadges                  str
CarTitle                   str
CarSubTitle                str
CarAttentionGrabber        str
CarSpecs                   str
CarSeller                  str
CarSellerRating        float64
CarSellerLocation          str
Year                       str
Reg                        str
BodyType                   str
Miles                  float64
EngineVol              float64
EngineSize             float64
EngineSizeUnit             str
Transmission               str
FuelType                   str
NumOwner               float64
ULEZ                     int64
FullService              int64
PartService              int64
PartWarranty             int64
FullDealership           int64
FirstYearRoadTax         int64
BrandNew                 int64
FinanceAvailable         int64
Discounted               int64
dtype: object

#### Boolean Columns

Some of the `int64` columns only contain 0s or 1s, and thus can be converted to booleans. We check every `int64` column, and convert it to `bool` iff its min and max values are in $\{0,1\}$.

In [7]:
def check_and_change(col_name):
    col = df[col_name]

    if not pd.api.types.is_integer_dtype(col):
        return False
    
    if col.max() > 1 or col.min() < 0:
        return False
    
    df[col_name] = df[col_name].astype(bool)
    return True


for col_name in df.columns.values:
    if check_and_change(col_name):
        print(f"{col_name} changed from an integer column to a boolean column")

df.head()

ULEZ changed from an integer column to a boolean column
FullService changed from an integer column to a boolean column
PartService changed from an integer column to a boolean column
PartWarranty changed from an integer column to a boolean column
FullDealership changed from an integer column to a boolean column
FirstYearRoadTax changed from an integer column to a boolean column
BrandNew changed from an integer column to a boolean column
FinanceAvailable changed from an integer column to a boolean column
Discounted changed from an integer column to a boolean column


,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,NumOwner,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted
0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",Private seller,...,5.0,False,False,False,False,False,False,False,False,False
1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",Boss Motor Company Limited,...,NaN,False,False,False,False,False,False,False,False,False
2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",PREMIER GT LIMITED,...,3.0,False,False,False,False,False,False,False,False,False
3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",TOP 555 LIMITED,...,NaN,False,False,False,False,False,False,False,False,False
4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",TOP 555 LIMITED,...,NaN,False,False,False,False,False,False,False,False,False


9 columns were changed by this process:

1. `ULEZ`
2. `FullService`
3. `PartService`
4. `PartWarranty`
5. `FullDealership`
6. `FirstYearRoadTax`
7. `BrandNew`
8. `FinanceAvailable`
9. `Discounted`

#### Converting `NumOwner` to Integer

We convert the column `NumOwner` to integer.

In [8]:
df["NumOwner"] = df["NumOwner"].astype(pd.Int64Dtype())
df[["Make", "Model", "Variant", "NumOwner"]].head(20)

,Make,Model,Variant,NumOwner
0,AC,Cobra,NaN,5
1,AC,Cobra,NaN,<NA>
2,AC,Cobra,NaN,3
3,AC,Cobra,NaN,<NA>
4,AC,Cobra,NaN,<NA>
5,AC,Cobra,NaN,<NA>
6,AC,NaN,MK IV,<NA>
7,AC,NaN,MK IV CRS,5
8,Abarth,124 Spider,GT,2
9,Abarth,124 Spider,GT,2


#### Converting `Year` to Integer

Currently `Year` is a string column. From initial analysis, we know that this is because some entries into the column are erroneous and are misentries belonging to different columns such as `Miles` or `BodyType`. For now, we simply convert to integer, like `NumOwners` and coerce any non-numeric value into NaN. We will concern ourselves with the erroneous numeric entries later.

In [9]:
df["Year"] = pd.to_numeric(df["Year"], errors="coerce").astype(pd.Int64Dtype())
df[["Make", "Model", "Variant", "Year"]].tail(20)

,Make,Model,Variant,Year
818436,Volvo,XC90,h T8,<NA>
818437,Volvo,XC90,h T8,2022
818438,Volvo,XC90,h T8,<NA>
818439,Volvo,XC90,h T8,2022
818440,Volvo,XC90,h T8,<NA>
818441,Volvo,XC90,h T8,<NA>
818442,Westfield,Roadster,NaN,2020
818443,Westfield,SE,NaN,1999
818444,Westfield,SE,NaN,1991
818445,Westfield,SEight,NaN,1998


## 2.3. Data Recovery

In the initial analysis, we noted that some of the data in the columns are missing. We attempt to recover as much as this as possible automatically.

#### `Model` Recovery

Some rows are missing their `Model` value. Although initial analysis showed that in most cases `CarTitle` is simply a string concatenation of `Make` and `Model`, we cannot reliably recover the data using the column. Hence we take a different approach. There are a lot of duplicate (`Make`, `Model`) pairs in the dataset, many of these also have the same `CarTitle` value. Hence we do the following.

- Construct a map `CarTitle` $\rightarrow$ `Model` using the data in the database, by grouping by `CarTitle` and taking the modal `Model`.
- Use this to fill in missing `Model` entries where possible.
- Manually inspect what remains.

In [10]:
title_to_model = df.groupby("CarTitle")["Model"].apply(
    lambda s: s.mode().iloc[0] if not s.mode().empty else pd.NA
)
title_to_model

CarTitle
AC Cobra                        Cobra
AK Cobra                        Cobra
Abarth 124                 124 Spider
Abarth 124 SPIDER          124 Spider
Abarth 124 Spider          124 Spider
                            ...      
Westfield Seight               SEight
Westfield Sport                 Sport
Wolseley 1500                    1500
Yamaha YXZ 1000R            Yxz 1000r
Zimmer Golden Spirit    Golden Spirit
Name: Model, Length: 3946, dtype: str

In [11]:
mask = df["Model"].isna()
pre_map = len(df[mask])

df.loc[
    mask,
    "Model"
] = df.loc[mask, "CarTitle"].map(title_to_model)

mask = df["Model"].isna()
post_map = len(df[mask])

print(f"#NA models pre-map: {pre_map}")
print(f"#NA models post-map: {post_map}")

df[mask][["Make", "Model", "Variant", "CarTitle", "CarSubTitle"]]

#NA models pre-map: 303
#NA models post-map: 2


,Make,Model,Variant,CarTitle,CarSubTitle
5172,Alpine,NaN,Legende GT,Alpine A110 Legende GT,1.8 Turbo Legende GT DCT Euro 6 2dr
654124,Sunbeam,NaN,Unlisted,Sunbeam Alpine,NaN


Automated recovery was successful in all but 2 cases. However we can reasonably recover these models manually by inspecting the columns by eye.

In [12]:
df.loc[5172, "Model"] = "A110"
df.loc[654124, "Model"] = "Alpine"

df[mask][["Make", "Model", "Variant", "CarTitle", "CarSubTitle"]]

,Make,Model,Variant,CarTitle,CarSubTitle
5172,Alpine,A110,Legende GT,Alpine A110 Legende GT,1.8 Turbo Legende GT DCT Euro 6 2dr
654124,Sunbeam,Alpine,Unlisted,Sunbeam Alpine,NaN


In [13]:
print(f"Remaining N/A Model Entries: {df["Model"].isna().sum()}")

Remaining N/A Model Entries: 0


We have successfully recovered `Model` in all cases using a mixture of both automatic and manual techinques.

#### `Variant` (Non-) Recovery

In theory, `Variant` can also be recovered using the `CarTitle` and `CarSubTitle` columns. However, we choose not to here for three reasons:

1. Where the `Variant` value would appear in these columns is not consistent across records, thus an automated method would take a long time to develop since we would have to ensure we are extracting the correct substring.
2. `CarTitle` and `CarSubTitle` pairings may not have many duplicates in the table, meaning a method similar to the one we used to recover `Model` would be inefficient.
3. Initial analysis identified that many entries in the `Variant` column are erroneous (another reason why a `Model`-like method would be unlikely to work) and so it seems that this column is generally unreliable.

For these reasons, we will _not_ recover missing `Variant` entries nor correct the erroneous ones. This therefore means that, unfortunately, we are unable to use this as a feature in modelling.

#### `Year` Recovery

We recover missing `Year` data. However, first we check that all years are reasonable.

In [14]:
summary_cols = ["Make", "Model", "Variant", "Year", "Reg", "CarSpecs"]
df.nlargest(10, "Year")[summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
5,AC,Cobra,NaN,2022,22 reg,"2022 (22 reg), Convertible, 500 miles, Manual,..."
437,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
438,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
439,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
440,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
441,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 7 miles, 1.4L, 162BH..."
442,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 3 miles, 1.4L, 162BH..."
443,Abarth,595,F595,2022,NaN,"2022, Hatchback, 1.4L, 167PS, Manual, Petrol, ..."
445,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 10 miles, 1.4L, 167P..."
449,Abarth,595,F595,2022,72 reg,"2022 (72 reg), Hatchback, 10 miles, 1.4L, 167P..."


In [15]:
df.nsmallest(10, "Year")[summary_cols + ["CarSubTitle", "CarTitle"]]

,Make,Model,Variant,Year,Reg,CarSpecs,CarSubTitle,CarTitle
237167,Ford,Ranger,TDCi,32,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ...",RANGER WILDTRAK DOUB 3.2,Ford RANGER 2019.5MY
237824,Ford,Ranger,Wildtrak,32,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ...",RANGER WILDTRAK DOUB 3.2,Ford RANGER 2019.5MY
586693,Pontiac,Firebird,NaN,1079,NaN,"1079, 79,000 miles",NaN,Pontiac Firebird
5283,Alvis,Sixteen,NaN,1934,NaN,"1934, Saloon, 65,000 miles, 2.2L, Manual, Petr...",SF16.95 Light Saloon 4dr,Alvis Sixteen
5284,Alvis,Sixteen,NaN,1934,NaN,"1934, Saloon, 65,000 miles, 2.2L, Manual, Petr...",SF16.95 Light Saloon 4dr,Alvis Sixteen
612050,Rolls-Royce,20/25,Unlisted,1934,A reg,"1934 (A reg), Limousine, 18,200 miles, 3.6L, P...",3.6 1934 classic 20/25 4dr,Rolls-Royce 20/25
612051,Rolls-Royce,20/25,Unlisted,1935,NaN,"1935, Saloon, 77,000 miles, Manual, Petrol",Barker 4dr,Rolls-Royce 20/25
378539,MG,TA,NaN,1936,NaN,"1936, Convertible, 65,165 miles, 1.3L, Manual,...",1250 1.3 2dr,MG TA
378538,MG,TA,NaN,1937,NaN,"1937, Convertible, 13,920 miles, 1.3L, Manual,...",1250 1.3,MG TA
612455,Rover,P2,NaN,1938,NaN,"1938, Saloon, 70,000 miles, Manual, Petrol, 3 ...",NaN,Rover P2


Entries into the `Year` column range from 32 to 2022. Of course 2022 is reasonable, but 32 is not. 1079 is not reasonable either. Both 32 entries are Ford Rangers which have 2019 in their `CarTitle`, we can reasonably assume that this is the year related to the car.

In [16]:
df.loc[
    237167,
    "Year"
] = 2019

df.loc[
    237824,
    "Year"
] = 2019

df.loc[[237824, 237167], summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
237824,Ford,Ranger,Wildtrak,2019,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ..."
237167,Ford,Ranger,TDCi,2019,NaN,"32,375 miles, 3.2L, 200PS, Automatic, Diesel, ..."


We see if we can recover the Pontiac's `Year`, unfortunately this seems much more unlikely since it isn't mentioned in any of the listing columns. Pontiac Firebirds were produced from 1967 to 2002 so there is a reasonable chance that 1079 is a typo and was supposed to be 1979. Nonetheless, we check:

- The `Year` column of other Pontiac Firebirds in the dataset
- If 1079 appears in any of the other columns for this particular Firebird

In [17]:
df[
    (df["Make"] == "Pontiac")
    & (df["Model"] == "Firebird")
].sort_values("Year")[summary_cols + ["CarPrice"]]

,Make,Model,Variant,Year,Reg,CarSpecs,CarPrice
586693,Pontiac,Firebird,NaN,1079,NaN,"1079, 79,000 miles",16500.0
586695,Pontiac,Firebird,NaN,1981,W reg,"1981 (W reg), Coupe, 44,333 miles, 4.9L, Autom...",34995.0
586694,Pontiac,Firebird,NaN,1990,G reg,"1990 (G reg), 77,000 miles",24995.0


In [18]:
df.select_dtypes(include="number").iloc[586693]

CarPrice           16500.0
CarSellerRating        4.0
Year                1079.0
Miles              79000.0
EngineVol             <NA>
EngineSize            <NA>
NumOwner              <NA>
Name: 586693, dtype: Float64

While it's most likely that 1979 is the intended year for the Pontiac Firebird, we cannot recover it with certainty, and thus we will replace it with NA.

In [19]:
df.loc[
    586693,
    "Year"
] = pd.NA

df.loc[[586693], summary_cols]

,Make,Model,Variant,Year,Reg,CarSpecs
586693,Pontiac,Firebird,NaN,<NA>,NaN,"1079, 79,000 miles"


We now inspect the rest of the entries with N/A `Year`. We know from initial analysis that `CarSpecs` will **not** contain the year. This is because we found that `CarSpecs` follows a fixed pattern which is never shuffled, so if a year is N/A or incorrect, it means that the correct year was simply omitted from `CarSpecs`. However, as we found with the Ford Rangers, these years may be present in the `CarTitle` or `CarSubTitle` columns.

In [20]:
recovery_cols = ["Make", "Model", "Variant", "Year", "CarTitle", "CarSubTitle", "BrandNew"]

print(f"{len(df[df["Year"].isna()])} rows with Year = NA")

df.loc[
    df["Year"].isna(),
    recovery_cols
].sample(100)

48081 rows with Year = NA


,Make,Model,Variant,Year,CarTitle,CarSubTitle,BrandNew
614189,SEAT,Arona,SE Technology,<NA>,SEAT Arona,1.0 TSI SE Technology (s/s) 5dr,True
297735,Jeep,Compass,GSE T4,<NA>,Jeep Compass,1.3 GSE T4 11.4kWh Upland Auto 4WD Euro 6 (s/s...,True
308420,Kia,Picanto,DPi,<NA>,KIA Picanto,1.0 DPi X-Line S AMT Euro 6 (s/s) 5dr,True
62972,Audi,TTS,Black Edition,<NA>,Audi Tts Coupe,AUDI TTS Coupe Black Edition TFSI 320 PS S tro...,True
409929,Mazda,MX-5 RF,GT Sport Tech,<NA>,Mazda MX-5,RF 2.0 184ps GT Sport Tech / White Nappa Leath...,True
...,...,...,...,...,...,...,...
459398,Mercedes-Benz,C Class,MHEV,<NA>,Mercedes-Benz C Class,2.0 C300h MHEV AMG Line (Premium Plus) G-Troni...,True
711392,Vauxhall,Corsa,SE Premium,<NA>,Vauxhall Corsa,1.2 SE Premium 5dr,True
758831,Volkswagen,Golf,Style,<NA>,Volkswagen Golf,1.5 TSI Style Euro 6 (s/s) 5dr,True
364703,Lexus,ES 300h,ES Premium Edition,<NA>,Lexus ES 300h,2.5 ES Premium Edition E-CVT (s/s) 4dr,True


There are ~48k rows which do not have a `Year` value. This is far many to check by hand, so we will leave them as NA. This only reflects a very small subset of all dataset records, so if Year seems to be quintessential in predicting price, removing these will not have a grave impact on the project. However, we notice that a lot of them are "brand new".

In [21]:
df[
    df["Year"].isna()
    & df["BrandNew"]
][recovery_cols]

,Make,Model,Variant,Year,CarTitle,CarSubTitle,BrandNew
350,Abarth,595,Competizione,<NA>,Abarth 595,595 Competizione 1.4 Tjet 180hp 3dr,True
351,Abarth,595,Competizione,<NA>,Abarth 595,1.4 T-Jet Competizione Euro 6 3dr,True
444,Abarth,595,F595,<NA>,Abarth Abarth 500,F595 1.4 Tjet 165hp 3dr,True
446,Abarth,595,F595,<NA>,Abarth 595,1.4 T-Jet F595 Euro 6 3dr,True
447,Abarth,595,F595,<NA>,Abarth Abarth 500,F595 1.4 Tjet 165hp 3dr,True
...,...,...,...,...,...,...,...
818435,Volvo,XC90,h T8,<NA>,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Bright Auto ...,True
818436,Volvo,XC90,h T8,<NA>,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Bright Auto ...,True
818438,Volvo,XC90,h T8,<NA>,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Bright SUV 5...,True
818440,Volvo,XC90,h T8,<NA>,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Dark SUV 5dr...,True


Exactly 47,929 of the rows missing a year are brand new, meaning only 152 are not brand new.

In [22]:
df.groupby("BrandNew")["Year"].agg(
    ["count", "median", "mean", "min", "max"]
)

,count,median,mean,min,max
BrandNew,,,,,
False,764095,2017.0,2015.989907,1934,2022
True,6280,2022.0,2021.812898,2021,2022


In addition, all brand new cars in the data set are from either 2021 or 2022. However, this is not enough to reliably recover the missing `Year` for these entries, without some form of guess work. Thus, we shall leave `Year` alone for now. However, this information could be crucial for retaining some data if we discover that `Year` is quintessential for `CarPrice` prediction.

#### `CarSellerLocation` Standardisation & Recovery

We standardise `CarSellerLocation` to title case, and recover any erroneous cases.

In [23]:
df["CarSellerLocation"] = df["CarSellerLocation"].str.title()

df["CarSellerLocation"].unique()

<ArrowStringArray>
[       'Watford',        'Chesham',     'Pulborough',         'Oakham',
  'Milton Keynes',         'Forres',      'Willesden',       'Alcester',
     'Nottingham',     'Portsmouth',
 ...
     'Newtonmore',    'Nr Amersham',       'Tregaron',      'Stromness',
 'Bridge Of Weir',            '3.1',       'Begbroke',  'Little Hadham',
        'Ibstock',   'Isle Of Skye']
Length: 1939, dtype: str

In [24]:
df[df["CarSellerLocation"] == "3.1"]["CarSeller"].unique()

<ArrowStringArray>
['Parkway Derby']
Length: 1, dtype: str

It seems that some instances of `CarSellerLocation` have value 3.1, which is certainly wrong. All rows with this error all have `CarSeller` "Parkway Derby", which is in Derby.

In [25]:
df.loc[
    df["CarSellerLocation"] == "3.1",
    "CarSellerLocation"
] = "Derby"

df[df["CarSellerLocation"] == "3.1"]["CarSeller"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

#### `Reg` Standardisation and Recovery

Every element of the `Reg` column has the form `__ reg` (or is NaN). We simply replace every entry with the first word of the string. We are unable to change the data type since, whilst newer reg plates are numerical, letters were used pre-2001. Thus the `Reg` column must remain a column of strings.

In [26]:
df["Reg"] = df["Reg"].str.split(" ").str[0]
df["Reg"].sample(100)

405660    NaN
730660     68
628393     18
249405     16
10468      69
         ... 
522841     66
101974    NaN
788387     69
257055     18
419370     67
Name: Reg, Length: 100, dtype: object

In [27]:
df["Reg"].isna().sum()

np.int64(71020)

71,020 columns do not contain a registration plate.

In [28]:
check_df = df[
    df["Reg"].isna()
][["Reg", "CarTitle", "CarSubTitle", "CarSpecs"]].copy()

check_df.sample(20)

,Reg,CarTitle,CarSubTitle,CarSpecs
125536,NaN,BMW X5,3.0 40d MHT M Sport Auto xDrive Euro 6 (s/s) 5dr,"SUV, 3.0L, 340PS, Automatic, Diesel Hybrid, Fi..."
49764,NaN,Audi Q5,2.0 TFSI 45 S line S Tronic quattro Euro 6 (s/...,"SUV, 2.0L, 265PS, Automatic, Petrol, First yea..."
297600,NaN,Jeep Compass,1.5 GSE Night Eagle SUV 5dr Petrol DCT FWD Eur...,"SUV, 1.5L, 130PS, Automatic, Petrol, First yea..."
556944,NaN,Nissan Qashqai,1.3 MHEV DIG-T Tekna (s/s) 5dr,"SUV, 1.3L, 158PS, Manual, Petrol, First year r..."
594771,NaN,Renault Arkana,1.3 TCe MHEV RS Line EDC 2WD Euro 6 (s/s) 5dr,"SUV, 1.3L, 138PS, Automatic, Petrol, First yea..."
719598,NaN,Vauxhall Crossland,Crossland 1.2T 130ps Ultimate AT 5dr,"SUV, 1.2L, 130PS, Manual, Petrol, First year r..."
471911,NaN,Mercedes-Benz E Class,2.0 E220d AMG Line (Premium) Estate 5dr Diesel...,"2018, Estate, 30,500 miles, 2.0L, 194PS, Autom..."
182201,NaN,Ford Ecosport,1.0T EcoBoost ST-Line Euro 6 (s/s) 5dr,"SUV, 1.0L, 125PS, Manual, Petrol, First year r..."
188479,NaN,Ford Fiesta,1.0 125ps Trend MHEV 5dr,"2022, Hatchback, 1.0L, 125PS, Manual, Petrol H..."
305381,NaN,KIA Niro,64.8kWh 2 Auto 5dr,"SUV, 204PS, Automatic, Electric, First year ro..."


In [29]:
pattern = r"\b[A-Za-z0-9]+ reg\b"

title_contains = check_df["CarTitle"].str.contains(pattern, case=False, na=False).rename("TitleContains")
subtitle_contains = check_df["CarSubTitle"].str.contains(pattern, case=False, na=False).rename("SubtitleContains")
specs_contains = check_df["CarSpecs"].str.contains(pattern, case=False, na=False).rename("SpecsContains")

check_df[title_contains]

,Reg,CarTitle,CarSubTitle,CarSpecs


In [30]:
check_df[subtitle_contains]

,Reg,CarTitle,CarSubTitle,CarSpecs
411057,NaN,Mazda 2,"1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 ...","2022, Hatchback, 4,200 miles, 1.5L, 90PS, Manu..."
413068,NaN,Mazda 2,"1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 ...","2022, Hatchback, 4,200 miles, 1.5L, 90PS, Manu..."
413827,NaN,Mazda 2,"1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 ...","2022, Hatchback, 4,200 miles, 1.5L, 90PS, Manu..."
677555,NaN,Toyota Estima,"2008 Reg, 2.4L, 5 Door, Automatic, Petrol","2022, MPV, 60,000 miles, 2.4L, Automatic, Petr..."
749170,NaN,Volkswagen Golf,"GT TDI BMT 19"" PRETORIA ALLOY WHEELS NI REG FR...","2018, Hatchback, 59,931 miles, 1.6L, 115PS, Ma..."


In [31]:
check_df[specs_contains]

,Reg,CarTitle,CarSubTitle,CarSpecs


Some of the entries with missing registrations may contain a registration string in their `CarSubTitle` column.

In [32]:
check_df[subtitle_contains]["CarSubTitle"].values

<ArrowStringArray>
[                                                    '1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 REG, IMMEDIATE DELIVERY)',
                                                     '1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 REG, IMMEDIATE DELIVERY)',
                                                     '1.5 90PS SPORT 5 DOOR (STUNNING EX DEMO, 2022 REG, IMMEDIATE DELIVERY)',
                                                                                  '2008 Reg, 2.4L, 5 Door, Automatic, Petrol',
 'GT TDI BMT 19" PRETORIA ALLOY WHEELS NI REG FROM NEW FULL VW SERVICE HISTORY PARKING SENSORS SAT NAV PRIVACY GLASS 1.6 5dr']
Length: 5, dtype: str

The `CarSubTitle` entries do not prove useful in recovering. 4 of them have the form `<year> reg` which is invalid for the UK, and another has the form `NI reg` which is also invalid for the UK. Thus we leave the NaN `Reg` entries as is.

## 2.4. Calculating New Columns

#### Calculating `EngineBHP`

We introduce a new column into the dataframe which unifies the `EngineSize` and `EngineSizeUnit` columns by converting all engine powers into BHP.

In [33]:
df["EngineSizeUnit"].unique()

<ArrowStringArray>
['bhp', nan, 'ps']
Length: 3, dtype: str

There are exactly 2 units used: BHP and PS. They are related as follows: 

$$
    1 \text{ PS} = 0.98632 \text{ BHP}
$$

When calculating the new column `EngineBHP`, we set values to NaN if either `EngineSize` or `EngineSizeUnit` is NaN.

In [34]:
df["EngineBHP"] = np.nan

bhp = df["EngineSizeUnit"].str.lower().eq("bhp")
ps = df["EngineSizeUnit"].str.lower().eq("ps")

df.loc[bhp, "EngineBHP"] = df.loc[bhp, "EngineSize"]
df.loc[ps, "EngineBHP"] = 0.98632 * df.loc[ps, "EngineSize"]

df[["Make", "Model", "Variant", "EngineSize", "EngineSizeUnit", "EngineBHP"]].head(20)

,Make,Model,Variant,EngineSize,EngineSizeUnit,EngineBHP
0,AC,Cobra,NaN,225.0,bhp,225.0000
1,AC,Cobra,NaN,NaN,NaN,NaN
2,AC,Cobra,NaN,NaN,NaN,NaN
3,AC,Cobra,NaN,NaN,NaN,NaN
4,AC,Cobra,NaN,NaN,NaN,NaN
5,AC,Cobra,NaN,NaN,NaN,NaN
6,AC,Cobra,MK IV,NaN,NaN,NaN
7,AC,Cobra,MK IV CRS,225.0,bhp,225.0000
8,Abarth,124 Spider,GT,167.0,bhp,167.0000
9,Abarth,124 Spider,GT,167.0,bhp,167.0000


We also check that this column indeed contains reasonable values.

In [35]:
cols = ["Make", "Model", "Variant", "Year", "CarTitle", "CarSubTitle", "CarSpecs", "EngineSize", "EngineSizeUnit", "EngineBHP", "CarPrice"]

df.nsmallest(20, "EngineBHP")[cols]

,Make,Model,Variant,Year,CarTitle,CarSubTitle,CarSpecs,EngineSize,EngineSizeUnit,EngineBHP,CarPrice
260701,Hyundai,Ioniq,Premium SE,2018,Hyundai Ioniq,1.6 h-GDi Premium SE DCT Euro 6 (s/s) 5dr,"2018 (18 reg), Hatchback, 53,000 miles, 1.6L, ...",1.0,bhp,1.00000,17494.0
260915,Hyundai,Ioniq,SE,2017,Hyundai Ioniq,1.6 h-GDi SE DCT Euro 6 (s/s) 5dr,"2017 (67 reg), Hatchback, 23,570 miles, 1.6L, ...",1.0,bhp,1.00000,17094.0
261035,Hyundai,Ioniq,h-GDi,2017,Hyundai Ioniq,1.6 h-GDi SE DCT Euro 6 (s/s) 5dr,"2017 (67 reg), Hatchback, 23,570 miles, 1.6L, ...",1.0,bhp,1.00000,17094.0
261045,Hyundai,Ioniq,h-GDi,2018,Hyundai Ioniq,1.6 h-GDi Premium SE DCT Euro 6 (s/s) 5dr,"2018 (18 reg), Hatchback, 53,000 miles, 1.6L, ...",1.0,bhp,1.00000,17494.0
133339,Citroen,Ami,NaN,2022,Citroen AMI,5.5kWh Auto 2dr,"2022 (72 reg), Coupe, 200 miles, 8PS, Automati...",8.0,ps,7.89056,7990.0
133340,Citroen,Ami,NaN,<NA>,Citroen AMI,5.5kWh Auto 2dr,"Coupe, 8PS, Automatic, Electric, First year ro...",8.0,ps,7.89056,8045.0
133341,Citroen,Ami,NaN,<NA>,Citroen AMI,5.5kWh Auto 2dr,"Coupe, 8PS, Automatic, Electric, First year ro...",8.0,ps,7.89056,8095.0
133342,Citroen,Ami,NaN,2022,Citroen AMI,5.5kWh Auto 2dr,"2022, Coupe, 8PS, Automatic, Electric, First y...",8.0,ps,7.89056,8095.0
133343,Citroen,Ami,NaN,2022,Citroen AMI,5.5kWh Auto 2dr,"2022, Coupe, 8PS, Automatic, Electric, First y...",8.0,ps,7.89056,8095.0
133338,Citroen,Ami,NaN,<NA>,Citroen AMI,6kW Ami 5.5kWh 2dr Auto Electric Coupe,"Coupe, 8BHP, Automatic, Electric, First year r...",8.0,bhp,8.00000,7695.0


In [36]:
df.nlargest(10, "EngineBHP")[cols]

,Make,Model,Variant,Year,CarTitle,CarSubTitle,CarSpecs,EngineSize,EngineSizeUnit,EngineBHP,CarPrice
746696,Volkswagen,Caravelle,Executive,2009,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",2460.0,bhp,2460.00000,16400.0
746844,Volkswagen,Caravelle,TDI,2009,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",2460.0,bhp,2460.00000,16400.0
167101,Ferrari,Sf90 Spider,T V8,2021,Ferrari SF90,Spider 2dr Auto 4.0,"2021 (71 reg), Convertible, 1,390 miles, 4.0L,...",986.0,ps,972.51152,479000.0
167102,Ferrari,Sf90 Spider,T V8,2021,Ferrari SF90,Spider 2dr Auto 4.0,"2021 (71 reg), Convertible, 1,376 miles, 4.0L,...",986.0,ps,972.51152,489000.0
167103,Ferrari,Sf90 Spider,T V8,2022,Ferrari SF90 Spider,4.0T V8 7.9kWh F1 DCT 4WD Euro 6 (s/s) 2dr,"2022 (22 reg), Convertible, 856 miles, 4.0L, 9...",986.0,ps,972.51152,499990.0
167104,Ferrari,Sf90 Spider,T V8,2022,Ferrari SF90 Spider,4.0T V8 7.9kWh F1 DCT 4WD Euro 6 (s/s) 2dr,"2022 (22 reg), Convertible, 2,206 miles, 4.0L,...",986.0,ps,972.51152,509830.0
167105,Ferrari,Sf90 Spider,T V8,2022,Ferrari SF90 SPIDER,4.0T V8 7.9kWh F1 DCT 4WD Euro 6 (s/s) 2dr,"2022 (22 reg), Convertible, 2,725 miles, 4.0L,...",986.0,ps,972.51152,529950.0
167106,Ferrari,Sf90 Spider,T V8,2021,Ferrari SF90 Spider,4.0T V8 7.9kWh F1 DCT 4WD Euro 6 (s/s) 2dr,"2021 (71 reg), Convertible, 350 miles, 4.0L, 9...",986.0,ps,972.51152,529995.0
167107,Ferrari,Sf90 Spider,T V8,2022,Ferrari SF90 Spider,4.0T V8 7.9kWh F1 DCT 4WD Euro 6 (s/s) 2dr,"2022 (22 reg), Convertible, 90 miles, 4.0L, 98...",986.0,ps,972.51152,539995.0
167108,Ferrari,Sf90 Spider,T V8,2022,Ferrari SF90,Spider 2dr Auto 4.0,"2022 (22 reg), Convertible, 2,712 miles, 4.0L,...",986.0,ps,972.51152,549000.0


The Hyundai Ioniq and Tucson certainly have more than 1 and 11 BHP respectively. We look for other obvious outliers by looking at rows which have 5x more or less of the median of the BHP of their (Make, Model) pair.

In [37]:
group_counts = df.groupby(["Make", "Model"])["EngineBHP"].transform("count")
df_bhp = df[group_counts >= 10].copy()

group_median = df_bhp.groupby(["Make", "Model"])["EngineBHP"].transform("median")

group_median

8         167.6744
9         167.6744
10        167.6744
11        167.6744
12        167.6744
            ...   
818437    231.7852
818438    231.7852
818439    231.7852
818440    231.7852
818441    231.7852
Name: EngineBHP, Length: 815683, dtype: float64

In [38]:
df_bhp["BHPRatio"] = df_bhp["EngineBHP"] / group_median

outliers = df_bhp[
    ((df_bhp["BHPRatio"] > 5)
    | (df_bhp["BHPRatio"] < 0.2))
    & (df["EngineBHP"] > 11)
]

outliers

/tmp/ipykernel_112936/2671006709.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  outliers = df_bhp[


,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted,EngineBHP,BHPRatio
487045,Mercedes-Benz,G Class,GD,16995.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,88.0,0.164008
487046,Mercedes-Benz,G Class,GD,17500.0,NaN,Mercedes-Benz G Class,3.0 GE300 GD 3dr,Merc GD240 4WD 2.4 3dr Ex Army,"1987 (E reg), SUV, 18,411 miles, 3.0L, 88BHP, ...",Private seller,...,False,False,False,False,False,False,False,False,88.0,0.164008
487048,Mercedes-Benz,G Class,GD,18200.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,88.0,0.164008
487061,Mercedes-Benz,G Class,GE300,16995.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,88.0,0.164008
487062,Mercedes-Benz,G Class,GE300,17500.0,NaN,Mercedes-Benz G Class,3.0 GE300 GD 3dr,Merc GD240 4WD 2.4 3dr Ex Army,"1987 (E reg), SUV, 18,411 miles, 3.0L, 88BHP, ...",Private seller,...,False,False,False,False,False,False,False,False,88.0,0.164008
487063,Mercedes-Benz,G Class,GE300,18200.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,88.0,0.164008
746696,Volkswagen,Caravelle,Executive,16400.0,NaN,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,NaN,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",Private seller,...,True,False,False,False,False,False,False,False,2460.0,16.627464
746844,Volkswagen,Caravelle,TDI,16400.0,NaN,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,NaN,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",Private seller,...,True,False,False,False,False,False,False,False,2460.0,16.627464


There is not enough data to recover the correct `EngineBHP` reliably, so we just replace them with NaN.

In [39]:
df.loc[
    outliers.index,
    "EngineBHP"
] = np.nan

df.loc[
    outliers.index,
    "EngineSize"
] = np.nan

df.iloc[outliers.index]

,Make,Model,Variant,CarPrice,CarBadges,CarTitle,CarSubTitle,CarAttentionGrabber,CarSpecs,CarSeller,...,ULEZ,FullService,PartService,PartWarranty,FullDealership,FirstYearRoadTax,BrandNew,FinanceAvailable,Discounted,EngineBHP
487045,Mercedes-Benz,G Class,GD,16995.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,False,NaN
487046,Mercedes-Benz,G Class,GD,17500.0,NaN,Mercedes-Benz G Class,3.0 GE300 GD 3dr,Merc GD240 4WD 2.4 3dr Ex Army,"1987 (E reg), SUV, 18,411 miles, 3.0L, 88BHP, ...",Private seller,...,False,False,False,False,False,False,False,False,False,NaN
487048,Mercedes-Benz,G Class,GD,18200.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,False,NaN
487061,Mercedes-Benz,G Class,GE300,16995.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,False,NaN
487062,Mercedes-Benz,G Class,GE300,17500.0,NaN,Mercedes-Benz G Class,3.0 GE300 GD 3dr,Merc GD240 4WD 2.4 3dr Ex Army,"1987 (E reg), SUV, 18,411 miles, 3.0L, 88BHP, ...",Private seller,...,False,False,False,False,False,False,False,False,False,NaN
487063,Mercedes-Benz,G Class,GE300,18200.0,'',Mercedes-Benz G Class,3.0 GE300 GD 3dr,07768050309 for Viewings,"1987 (E reg), SUV, 84,800 miles, 3.0L, 88BHP, ...",Hollins Hill Cars,...,False,False,False,False,False,False,False,False,False,NaN
746696,Volkswagen,Caravelle,Executive,16400.0,NaN,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,NaN,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",Private seller,...,False,True,False,False,False,False,False,False,False,NaN
746844,Volkswagen,Caravelle,TDI,16400.0,NaN,Volkswagen Caravelle,2.5 TDI Executive Tiptronic SWB Euro 3 5dr,NaN,"2009 (09 reg), MPV, 158,362 miles, 2.5L, 2460B...",Private seller,...,False,True,False,False,False,False,False,False,False,NaN


## 2.5. Outputting Cleaned Files

We output cleaned data using `parquet` files. We will create 2 files. The first (`data/clean_car_adverts.parquet`) will contain the dataframe as it stands at this current point, i.e. after the cleaning steps above.

In [40]:
df.to_parquet("../data/clean_car_adverts.parquet")

The second file (`data/car_data.parquet`) will contain a stripped down version of the dataframe which only contains data which we wish to use for prediction. We remove any columns which are either redundant or related to the car advert:

- `CarTitle`
- `CarSubtitle`
- `CarSpecs`
- `CarBadges`
- `CarAttentionGrabber`
- `CarSeller`
- `CarSellerRating`
- `CarSellerLocation`
- `FullService`
- `PartService`
- `PartWarranty`
- `FullDealership`
- `FirstYearRoadTax`
- `FinanceAvailable`
- `Discounted`
- `EngineSize`
- `EngineSizeUnit`

In [41]:
small_df = df.drop(columns=[
    "CarTitle",
    "CarSubTitle",
    "CarSpecs",
    "CarBadges",
    "CarAttentionGrabber",
    "CarSeller",
    "CarSellerRating",
    "CarSellerLocation",
    "FullService",
    "PartService",
    "PartWarranty",
    "FullDealership",
    "FirstYearRoadTax",
    "FinanceAvailable",
    "Discounted",
    "EngineSize",
    "EngineSizeUnit"
])

small_df.head()

,Make,Model,Variant,CarPrice,Year,Reg,BodyType,Miles,EngineVol,Transmission,FuelType,NumOwner,ULEZ,BrandNew,EngineBHP
0,AC,Cobra,NaN,89995.0,2001,X,convertible,14400.0,4.9,manual,petrol,5,False,False,225.0
1,AC,Cobra,NaN,92500.0,2019,T,convertible,650.0,NaN,manual,petrol,<NA>,False,False,NaN
2,AC,Cobra,NaN,109995.0,2000,X,convertible,21600.0,3.5,manual,petrol,3,False,False,NaN
3,AC,Cobra,NaN,124950.0,1989,F,convertible,2750.0,NaN,manual,petrol,<NA>,False,False,NaN
4,AC,Cobra,NaN,124950.0,1989,E,convertible,15142.0,5.0,manual,petrol,<NA>,False,False,NaN


In [42]:
small_df.to_parquet("../data/car_data.parquet")